etcdserver/server.go

```diff
@@ -775,9 +775,57 @@ func (s *EtcdServer) triggerSnapshot(ep *etcdProgress) {
 	ep.snapi = ep.appliedi
 }
 
-// Stop stops the server gracefully, and shuts down the running goroutine.
-// Stop should be called after a Start(s), otherwise it will block forever.
-func (s *EtcdServer) Stop() {
+// TransferLeadership transfers the leader to the given transferee.
+func (s *EtcdServer) TransferLeadership(ctx context.Context, lead, transferee uint64) error {
+	now := time.Now()
+	interval := time.Duration(s.Cfg.ElectionTicks) * time.Millisecond
+
+	donec, stopc := make(chan struct{}), make(chan struct{})
+	go func() {
+		defer close(donec)
+
+		plog.Infof("transferring leadership from %s to %s", types.ID(lead), types.ID(transferee))
+		s.r.TransferLeadership(ctx, lead, transferee)
+		for s.Lead() != transferee {
+			select {
+			case <-stopc:
+				return
+			case <-time.After(interval):
+			}
+		}
+	}()
+
+	select {
+	case <-donec:
+	case <-ctx.Done(): // time out
+		close(stopc)
+		return ErrTimeoutLeaderTransfer
+	}
+
+	// TODO: drain all requests, or drop all messages to the old leader
+
+	plog.Infof("finished leader transfer to %s (took %v)", types.ID(transferee), time.Since(now))
+	return nil
+}
+
+func (s *EtcdServer) stopServer(leaderTransfer bool) {
+	if leaderTransfer {
+		if s.cluster != nil && len(s.cluster.MemberIDs()) > 1 && uint64(s.ID()) == s.Lead() {
```

Comment:
Even the choosing transferee? I think it's good to have them separate in case we want to make this available to clients with manually chosen transferee?